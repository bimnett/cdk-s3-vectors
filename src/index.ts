export * from './s3-vectors/bucket';
export * from './s3-vectors/index';
export * from './s3-vectors/knowledge-base';