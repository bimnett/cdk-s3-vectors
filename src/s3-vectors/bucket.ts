import * as custom_resources from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';

export interface BucketProps {
  /**
    * The bucket where the vector data will be stored.
    */
  readonly bucketName: string;

  /**
    * The encryption configuration for the S3 Vector bucket.
    */
  readonly encryptionConfiguration?: EncryptionConfiguration;

  /**
    * The AWS region where the resource will be created.
    */
  readonly region: string;
}

export interface EncryptionConfiguration {
  /**
    * The server-side encryption type. Must be 'aws:kms'.
    */
  readonly sseType: 'aws:kms';

  /**
    * The KMS key ARN to use for encryption.
    */
  readonly kmsKeyArn: string;
}


export class Bucket extends Construct {
  /**
    * The bucket where the vector data will be stored.
    */
  public readonly bucketName: string;

  /**
    * The AWS region where the resource will be created.
    */
  public readonly region: string;

  /**
    * @summary Creates a new Bucket construct for S3 Vectors.
    * @param {cdk.App} scope - Represents the scope for all resources.
    * @param {string} id - Scope-unique id.
    * @param {IndexProps} props - User provided props for the construct.
    * @access public
    */

  constructor(scope: Construct, id: string, props: BucketProps) {
    super(scope, id);
    this.bucketName = props.bucketName;
    this.region = props.region;

    const createParams: { [key: string]: any } = {
      vectorBucketName: props.bucketName,
    };

    if (props.encryptionConfiguration) {
      createParams.encryptionConfiguration = props.encryptionConfiguration;
    }

    const physicalResourceId = custom_resources.PhysicalResourceId.of(
      [
        props.bucketName,
        props.region,
        JSON.stringify(props.encryptionConfiguration || {}),
      ].join('-'),
    );

    const onCreate: custom_resources.AwsSdkCall = {
      service: 'S3Vectors',
      action: 'createVectorBucket',
      parameters: createParams,
      physicalResourceId: physicalResourceId,
      region: props.region,
    };

    const onDelete: custom_resources.AwsSdkCall = {
      service: 'S3Vectors',
      action: 'deleteVectorBucket',
      parameters: { vectorBucketName: props.bucketName },
      region: props.region,
    };

    new custom_resources.AwsCustomResource(this, 'S3VectorStoreCustomResource', {
      onCreate: onCreate,
      onDelete: onDelete,
      policy: custom_resources.AwsCustomResourcePolicy.fromSdkCalls({
        resources: custom_resources.AwsCustomResourcePolicy.ANY_RESOURCE,
      }),
    });
  }
}