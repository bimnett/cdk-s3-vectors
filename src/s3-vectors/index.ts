import * as iam from 'aws-cdk-lib/aws-iam';
import { IGrantable } from 'aws-cdk-lib/aws-iam';
import * as custom_resources from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';
import { Bucket } from './bucket';

export interface IndexProps {
  /**
    * The S3 Vector Bucket construct where the index will be created.
    */
  readonly bucket: Bucket;

  /**
    * The name of the vector index.
    */
  readonly indexName: string;

  /**
    * The data type of the vectors in the index.
    */
  readonly dataType: 'float32';

  /**
    * The number of dimensions for the vectors in the index.
    */
  readonly dimension: number;

  /**
    * The distance metric to use for calculating similarity.
    */
  readonly distanceMetric: 'euclidean' | 'cosine';

  /**
    * Optional metadata configuration for the index.
    */
  readonly metadataConfiguration?: MetadataConfiguration;
}

export interface MetadataConfiguration {
  /**
    * A list of keys for metadata fields that should not be filterable.
    */
  readonly nonFilterableMetadataKeys: string[];
}

export class Index extends Construct {
  /**
    * The ARN (Amazon Resource Name) of the created S3 Vector index.
    */
  public readonly indexArn: string;

  /**
   * The HTTPS endpoint for the S3 Vector index, used for making API calls.
   */
  public readonly indexEndpoint: string;

  /**
    * @summary Creates a new Index construct for S3 Vectors.
    * @param {cdk.App} scope - Represents the scope for all resources.
    * @param {string} id - Scope-unique id.
    * @param {IndexProps} props - User provided props for the construct.
    * @access public
    */

  constructor(scope: Construct, id: string, props: IndexProps) {
    super(scope, id);
    if (props.dimension < 1 || props.dimension > 4096) {
      throw new Error('Dimension must be between 1 and 4096.');
    }

    const createParams: { [key: string]: any } = {
      vectorBucketName: props.bucket.bucketName,
      indexName: props.indexName,
      dataType: props.dataType,
      dimension: props.dimension,
      distanceMetric: props.distanceMetric,
    };

    if (props.metadataConfiguration) {
      createParams.metadataConfiguration = props.metadataConfiguration;
    }

    // Helper function to create a stable string representation of the metadata object
    const getStableMetadataString = (metadata?: MetadataConfiguration) => {
      if (!metadata || !metadata.nonFilterableMetadataKeys) {
        return '';
      }
      const sortedKeys = [...metadata.nonFilterableMetadataKeys].sort();
      return JSON.stringify({ nonFilterableMetadataKeys: sortedKeys });
    };

    const physicalResourceId = custom_resources.PhysicalResourceId.of(
      [
        props.indexName,
        props.dimension,
        props.dataType,
        props.distanceMetric,
        getStableMetadataString(props.metadataConfiguration),
      ].join('-'),
    );

    const onCreate: custom_resources.AwsSdkCall = {
      service: 'S3Vectors',
      action: 'createIndex',
      parameters: createParams,
      physicalResourceId: physicalResourceId,
      region: props.bucket.region,
    };

    const onDelete: custom_resources.AwsSdkCall = {
      service: 'S3Vectors',
      action: 'deleteIndex',
      parameters: {
        vectorBucketName: props.bucket.bucketName,
        indexName: props.indexName,
      },
      region: props.bucket.region,
    };

    const customResource = new custom_resources.AwsCustomResource(this, 'S3VectorIndexCustomResource', {
      onCreate: onCreate,
      onDelete: onDelete,
      policy: custom_resources.AwsCustomResourcePolicy.fromSdkCalls({
        resources: custom_resources.AwsCustomResourcePolicy.ANY_RESOURCE,
      }),
    });

    this.indexArn = customResource.getResponseField('IndexArn').toString();
    this.indexEndpoint = customResource.getResponseField('IndexEndpoint').toString();
    // Ensure the index is created only after the bucket has been created.
    customResource.node.addDependency(props.bucket);
  }

  /**
    * Grants write permissions (add/delete vectors) to the index.
    * @param grantee The principal to grant permissions to.
    */
  public grantWrite(grantee: IGrantable): void {
    grantee.grantPrincipal.addToPrincipalPolicy(new iam.PolicyStatement({
      actions: [
        'S3Vectors:addVectors',
        'S3Vectors:deleteVectors',
      ],
      resources: [this.indexArn],
    }));
  }
}

export { Bucket };
