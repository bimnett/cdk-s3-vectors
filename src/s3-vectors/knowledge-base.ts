import * as iam from 'aws-cdk-lib/aws-iam';
import * as custom_resources from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';
import { Bucket, BucketProps } from './bucket';
import { Index, IndexProps } from './index';

export interface S3VectorKnowledgeBaseProps {
  /**
    * The name of the knowledge base.
    */
  readonly knowledgeBaseName: string;

  /**
    * A description of the knowledge base.
    */
  readonly description?: string;

  /**
    * The Amazon Resource Name (ARN) of the model used to create vector embeddings for the knowledge base.
    */
  readonly embeddingModelArn: string;

  /**
    * User provided props for the S3 vectors bucket construct.
    */
  readonly bucketProps: BucketProps;

  /**
    * User provided props for the S3 vectors index construct.
    */
  readonly indexProps: IndexProps;
}

export class S3VectorKnowledgeBase extends Construct {
  /**
    * The ID of the knowledge base.
    */
  public readonly knowledgeBaseId: string;

  /**
    * The Amazon Resource Name (ARN) of the knowledge base.
    */
  public readonly knowledgeBaseArn: string;


  /**
    * @summary Creates a new Bedrock knowledge base construct with S3 Vectors as the vector store.
    * @param {cdk.App} scope - Represents the scope for all resources.
    * @param {string} id - Scope-unique id.
    * @param {IndexProps} props - User provided props for the construct.
    * @access public
    */


  constructor(scope: Construct, id: string, props: S3VectorKnowledgeBaseProps) {
    super(scope, id);

    const vectorBucket = new Bucket(this, 'S3VectorBucket', props.bucketProps);

    const vectorIndex = new Index(this, 'S3VectorIndex', {
      ...props.indexProps,
      bucket: vectorBucket,
    });

    const knowledgeBaseRole = new iam.Role(this, 'KnowledgeBaseRole', {
      assumedBy: new iam.ServicePrincipal('bedrock.amazonaws.com'),
    });

    knowledgeBaseRole.addToPolicy(new iam.PolicyStatement({
      actions: ['s3:GetObject', 's3:ListBucket'],
      resources: [vectorBucket.bucketArn, `${vectorBucket.bucketArn}/*`],
    }));
    knowledgeBaseRole.addToPolicy(new iam.PolicyStatement({
      actions: ['s3vectors:addVectors', 's3vectors:deleteVectors'],
      resources: [vectorIndex.indexArn],
    }));
    knowledgeBaseRole.addToPolicy(new iam.PolicyStatement({
      actions: ['bedrock:InvokeModel'],
      resources: [props.embeddingModelArn],
    }));

    const onCreate: custom_resources.AwsSdkCall = {
      service: 'BedrockAgent',
      action: 'createKnowledgeBase',
      parameters: {
        name: props.knowledgeBaseName,
        description: props.description,
        roleArn: knowledgeBaseRole.roleArn,
        knowledgeBaseConfiguration: {
          type: 'VECTOR',
          vectorKnowledgeBaseConfiguration: {
            embeddingModelArn: props.embeddingModelArn,
          },
        },
        storageConfiguration: {
          type: 'S3_VECTORS',
          s3VectorsConfiguration: {
            vectorBucketArn: vectorBucket.bucketArn,
            indexArn: vectorIndex.indexArn,
            indexName: vectorIndex.indexName,
          },
        },
      },
      physicalResourceId: custom_resources.PhysicalResourceId.of(props.knowledgeBaseName),
      outputPaths: [
        'knowledgeBase.knowledgeBaseId',
        'knowledgeBase.knowledgeBaseArn',
      ],
    };

    const onDelete: custom_resources.AwsSdkCall = {
      service: 'BedrockAgent',
      action: 'deleteKnowledgeBase',
      parameters: { knowledgeBaseId: '{{knowledgeBase.knowledgeBaseId}}' },
    };

    const knowledgeBaseResource = new custom_resources.AwsCustomResource(this, 'KnowledgeBaseCustomResource', {
      onCreate: onCreate,
      onDelete: onDelete,
      policy: custom_resources.AwsCustomResourcePolicy.fromSdkCalls({ resources: custom_resources.AwsCustomResourcePolicy.ANY_RESOURCE }),
    });

    this.knowledgeBaseId = knowledgeBaseResource.getResponseField('knowledgeBase.knowledgeBaseId').toString();
    this.knowledgeBaseArn = knowledgeBaseResource.getResponseField('knowledgeBase.knowledgeBaseArn').toString();

    // Ensure the knowledge base is created after the role, bucket, and index
    knowledgeBaseResource.node.addDependency(knowledgeBaseRole);
    knowledgeBaseResource.node.addDependency(vectorBucket);
    knowledgeBaseResource.node.addDependency(vectorIndex);
  }
}