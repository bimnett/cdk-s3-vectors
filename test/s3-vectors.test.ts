import { App, Stack } from 'aws-cdk-lib';
import { Template, Match } from 'aws-cdk-lib/assertions';
import { Role, ServicePrincipal, IGrantable } from 'aws-cdk-lib/aws-iam';
import { Bucket, Index } from '../src/s3-vectors';

// A helper construct to test the grantWrite method
class TestGrantee extends Stack implements IGrantable {
  public readonly grantPrincipal: Role;
  constructor(scope: App, id: string) {
    super(scope, id);
    this.grantPrincipal = new Role(this, 'TestRole', {
      assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
    });
  }
}

// ---- Bucket Construct Tests ----
describe('Bucket Construct', () => {
  let app: App;
  let stack: Stack;
  let template: Template;

  beforeEach(() => {
    app = new App();
    stack = new Stack(app, 'TestStack');
    new Bucket(stack, 'MyVectorBucket', {
      bucketName: 'my-test-bucket',
      region: 'us-east-1',
    });
    template = Template.fromStack(stack);
  });

  test('Creates a Custom Resource for the Bucket', () => {
    template.hasResourceProperties('Custom::AWS', {
      ServiceToken: Match.anyValue(),
      Create: Match.serializedJson({
        service: 'S3Vectors',
        action: 'createVectorBucket',
        parameters: Match.objectLike({
          vectorBucketName: 'my-test-bucket',
        }),
        physicalResourceId: Match.anyValue(),
        region: Match.anyValue(),
      }),
      Delete: Match.serializedJson({
        service: 'S3Vectors',
        action: 'deleteVectorBucket',
        parameters: Match.objectLike({
          vectorBucketName: 'my-test-bucket',
        }),
        region: Match.anyValue(),
      }),
    });
  });

  test('Bucket Snapshot Test', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});

// ---- Index Construct Tests ----
describe('Index Construct', () => {
  let app: App;
  let stack: Stack;
  let bucket: Bucket;
  let granteeStack: TestGrantee;

  beforeEach(() => {
    app = new App();
    stack = new Stack(app, 'TestStack');
    bucket = new Bucket(stack, 'MyVectorBucket', {
      bucketName: 'my-test-bucket',
      region: 'us-east-1',
    });
    granteeStack = new TestGrantee(app, 'TestGranteeStack');
  });

  test('Index Creates a Custom Resource with correct props', () => {
    new Index(stack, 'MyVectorIndex', {
      bucket: bucket,
      indexName: 'my-test-index',
      dataType: 'float32',
      dimension: 1536,
      distanceMetric: 'cosine',
    });
    const template = Template.fromStack(stack);
    template.hasResourceProperties('Custom::AWS', {
      ServiceToken: Match.anyValue(),
      Create: Match.serializedJson({
        service: 'S3Vectors',
        action: 'createIndex',
        parameters: Match.objectLike({
          vectorBucketName: 'my-test-bucket',
          indexName: 'my-test-index',
          dataType: 'float32',
          dimension: 1536,
          distanceMetric: 'cosine',
        }),
        physicalResourceId: Match.anyValue(),
        region: Match.anyValue(),
      }),
      Delete: Match.serializedJson({
        service: 'S3Vectors',
        action: 'deleteIndex',
        parameters: Match.objectLike({
          vectorBucketName: 'my-test-bucket',
          indexName: 'my-test-index',
        }),
        region: Match.anyValue(),
      }),
    });
  });

  test('Index has a dependency on the Bucket', () => {
    new Index(stack, 'MyVectorIndex', {
      bucket: bucket,
      indexName: 'my-test-index',
      dataType: 'float32',
      dimension: 1536,
      distanceMetric: 'cosine',
    });
    const template = Template.fromStack(stack);

    const bucketResource = template.findResources('Custom::AWS', {
      Properties: Match.objectLike({
        Create: Match.serializedJson({
          service: 'S3Vectors',
          action: 'createVectorBucket',
          parameters: { vectorBucketName: 'my-test-bucket' },
          physicalResourceId: Match.anyValue(),
          region: Match.anyValue(),
        }),
      }),
    });
    const bucketLogicalId = Object.keys(bucketResource)[0];

    template.hasResource('Custom::AWS', {
      Properties: Match.objectLike({
        Create: Match.serializedJson({
          service: 'S3Vectors',
          action: 'createIndex',
          parameters: Match.objectLike({
            vectorBucketName: 'my-test-bucket',
            indexName: 'my-test-index',
          }),
          physicalResourceId: Match.anyValue(),
          region: Match.anyValue(),
        }),
      }),
      DependsOn: Match.arrayWith([bucketLogicalId]),
    });
  });

  test('grantWrite adds correct policy to grantee', () => {
    const index = new Index(stack, 'MyVectorIndex', {
      bucket: bucket,
      indexName: 'my-test-index',
      dataType: 'float32',
      dimension: 1536,
      distanceMetric: 'cosine',
    });

    index.grantWrite(granteeStack);

    const granteeTemplate = Template.fromStack(granteeStack);

    granteeTemplate.hasResourceProperties('AWS::IAM::Policy', {
      PolicyDocument: Match.objectLike({
        Statement: Match.arrayWith([
          Match.objectLike({
            Action: ['S3Vectors:addVectors', 'S3Vectors:deleteVectors'],
            Resource: Match.objectLike({
              'Fn::ImportValue': Match.anyValue(),
            }),
          }),
        ]),
      }),
    });
  });

  test('Index validation throws for invalid dimension', () => {
    expect(() => {
      new Index(stack, 'InvalidIndex', {
        bucket: bucket,
        indexName: 'invalid-index',
        dataType: 'float32',
        dimension: 0,
        distanceMetric: 'cosine',
      });
    }).toThrow('Dimension must be between 1 and 4096.');
  });
});