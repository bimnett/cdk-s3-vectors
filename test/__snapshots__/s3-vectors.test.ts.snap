// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Bucket Construct Bucket Snapshot Test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "DependsOn": [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "11d371859e49f3b1a49896583254890801a40beada56e4c84409a179a221e2e2.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyVectorBucketS3VectorStoreCustomResourceCustomResourcePolicy7A53AB6E": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3vectors:CreateVectorBucket",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "s3vectors:DeleteVectorBucket",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyVectorBucketS3VectorStoreCustomResourceCustomResourcePolicy7A53AB6E",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "MyVectorBucketS3VectorStoreCustomResourceD70E6278": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "MyVectorBucketS3VectorStoreCustomResourceCustomResourcePolicy7A53AB6E",
      ],
      "Properties": {
        "Create": "{"service":"S3Vectors","action":"createVectorBucket","parameters":{"vectorBucketName":"my-test-bucket"},"physicalResourceId":{"id":"my-test-bucket-us-east-1-{}"},"region":"us-east-1"}",
        "Delete": "{"service":"S3Vectors","action":"deleteVectorBucket","parameters":{"vectorBucketName":"my-test-bucket"},"region":"us-east-1"}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
